---
title: "Recherche d'homologues"
author: "Gustavo Magaña López"
date: "20/11/2020"
output:
  pdf_document: default
  html_document: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r imports, message=FALSE, warning=FALSE, tidy=TRUE}
library(msa) # Paquet pour faire alignements multiples
library(aphid) # Paquet contenant les matrices de substitution
data("substitution") # Substitution matrices
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

```{r libraries}
#library(BSgenome.Scerevisiae.UCSC.sacCer3)
#c("backports", "cpp11", "dbplyr", "future", "lava", "renv")
```


When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r import.data, message=FALSE, warning=FALSE, tidy=TRUE}
filepath <- "data/struct/seqComRDelta-Blast.fasta"
mySeqs <- readAAStringSet(filepath, format = "fasta")
```

```{r susbs}
names(substitution)
```
## Including Plots

You can also embed plots, for example:
 msaClustalW(inputSeqs, cluster="default", gapOpening="default", 
                gapExtension="default", maxiters="default", 
                substitutionMatrix="default", type="default", 
                order=c("aligned", "input"), verbose=FALSE,
                help=FALSE, ...)

```{r pressure, echo=FALSE, message=FALSE, warning=FALSE, tidy=TRUE}
library(doParallel)
library(foreach)

f.algin.score <- function(seqs, go, ge, subm, gapval){ 
  # Compute score
  sum(
    # Compute conservation matrix
    msaConservationScore(
      # compute alignment using ClustalW
      msaClustalW(
        seqs, substitutionMatrix = subm, 
        gapOpening = go, gapExtension = ge,
        verbose = FALSE
      ), 
      substitutionMatrix = subm, gapVsGap = gapval
    )
  )
}

parallel.search <- function(
  filepath, subm, gapval, 
  go = seq(5, 15, 1), ge = seq(0.1, 1.5, 0.05), 
  nthreads=(detectCores()/2)
){
  cl <- parallel::makeCluster(nthreads, type = "FORK")
  doParallel::registerDoParallel(cl)
  
  #nreps <- length(ge) * length(go)
  #i.seqs <- iter(replicate(nreps, seqs, simplify = FALSE))
  grid.eval <- foreach(g_op = go, .combine = rbind) %dopar% {
    foreach(g_ext = ge, .combine = rbind) %dopar% {
      it.algn.score <- f.algin.score(
        filepath, g_op, g_ext, subm, gapval
      )
      data.frame(gop = g_op, gext = g_ext)
    }
  }
  stopCluster(cl) 
  return(grid.eval)
}

linear.search <- function(seqs, subm, gapval){
  retval <- foreach(g_op = seq(5, 15, 1), .combine = rbind) %do% {
    foreach(g_ext = seq(0.1, 1.5, 0.05), .combine = rbind) %do% {
      it.algn.score <- f.algin.score(seqs, g_op, g_ext, subm, gapval)
      data.frame(gop = g_op, gext = g_ext, score = it.algn.score)
    }
  }
  return(retval)
}
```

```{r auxilio}
#my.pam <- parallel.search(filepath, substitution$PAM150, 0)
```

```{r aber, message=FALSE, warning=FALSE, tidy=TRUE, cache=TRUE, results="hide"}
my.pam <- linear.search(mySeqs, substitution$PAM150, 0)
my.blos <- linear.search(mySeqs, substitution$BLOSUM60, 0)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r optimal}
linmax <- function(x){ x[x$score == max(x$score), ] }
```

```{r what, message=FALSE, warning=FALSE, tidy=TRUE, cache=TRUE, results="hide"}
gonnet <- linear.search(seqs = mySeqs, subm = substitution$IDENTITY, gapval = 0)
```

```{r plot}
library(ggplot2)

plot.align.search <- function(the.table, ttl="GridSearch"){
  gap.Open <- factor(the.table$gop)
  gap.Ext <- factor(the.table$gext)
  algn.score <- factor(the.table$score)
  qplot(x = gap.Open, y = gap.Ext, fill = algn.score, geom = 'tile', main = ttl)
}
```

## Blossum 
```{r blossum}
plot.align.search(my.blos, "woo")
linmax(my.blos)
```

## PAM
```{r pam}
plot.align.search(my.pam)
linmax(my.pam)
```


